core/vm/evm.go

```diff
@@ -160,6 +160,28 @@ func (evm *EVM) Interpreter() *EVMInterpreter {
 	return evm.interpreter
 }
 
+// SetBlockContext updates the block context of the EVM.
+func (evm *EVM) SetBlockContext(blockCtx BlockContext) {
+	evm.Context = blockCtx
+	evm.chainRules = evm.chainConfig.Rules(blockCtx.BlockNumber, blockCtx.Random != nil)
+}
+
+// SetTracer updates the EVM tracer and enables debugging. If the tracer is nil
+// any set tracer is removed and debugging is disabled.
+func (evm *EVM) SetTracer(tracer EVMLogger) {
+	if tracer == nil {
+		evm.Config.Debug = false
+		evm.Config.Tracer = nil
+		evm.interpreter.cfg.Debug = false
+		evm.interpreter.cfg.Tracer = nil
```

Comment:
This code assumes that `evm.interpreter` is non-nil, and furthermore that `interpreter.cfg` is non-nil. I'm not sure that's a safe assumption to make?

Reply:
In go-ethereum every instantiation is via vm.NewEVM which sets the interpreter. The EVM Config could theoretically be nil as well. However, there are no nil checks in the code in other methods like EVM.Call. Thus, I thought it is consistent not to check.

Reply:
Just noticed the EVMInterpreter holds a pointer to the EVM and a copy of the EVM.Config. So we could just drop the config from the interpreter and use it via the EVM instead. What do you thing @holiman ?